framework/src/samplers/Sampler.C

```diff
@@ -86,3 +112,33 @@ Sampler::setNumberOfRequiedRandomSeeds(const std::size_t & number)
 
   _generator.saveState();
 }
+
+void
+Sampler::setSampleNames(const std::vector<std::string> & names)
+{
+  _sample_names = names;
+
+  // Use assert because to check the size a getSamples call is required, which you don't
+  // want to do if you don't need it.
+  mooseAssert(getSamples().size() == _sample_names.size(),
+              "The number of sample names must match the number of samples returned.");
+}
+
+Sampler::Location
+Sampler::getLocation(unsigned int global_index)
+{
+  if (_offsets.empty())
+    reinit(getSamples());
+
+  std::vector<unsigned int>::const_iterator iter =
+      std::upper_bound(_offsets.begin(), _offsets.end(), global_index) - 1;
```

Comment:
`- 1` is dangerous, this will fail if the upper_bound returns an iterator pointing to the first item. Also, if no item is found you'll get the pointer to the end which means it's not found, however you'll point back to the real last item in that case. If all this is fine, it should probably be asserted and/or documented.